src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -1970,6 +1970,28 @@ internal MethodSymbol EnsureInlineArrayAsSpanExists(SyntaxNode syntaxNode, Named
                                                       diagnostics);
         }
 
+        internal NamedTypeSymbol EnsureInlineArrayTypeExists(SyntaxNode syntaxNode, SyntheticBoundNodeFactory factory, int arrayLength, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(Compilation.Assembly.RuntimeSupportsInlineArrayTypes);
+
+            string typeName = PrivateImplementationDetails.GetInlineArrayTypeName(arrayLength);
+            var privateImplClass = GetPrivateImplClass(syntaxNode, diagnostics);
+            var typeAdapter = privateImplClass.GetType(typeName);
+
+            if (typeAdapter is null)
+            {
+                var attributeConstructor = (MethodSymbol)factory.SpecialMember(SpecialMember.System_Runtime_CompilerServices_InlineArrayAttribute__ctor);
+                Debug.Assert(attributeConstructor is { });
+
+                var typeSymbol = new SynthesizedInlineArrayTypeSymbol(SourceModule, typeName, arrayLength, attributeConstructor);
+                privateImplClass.TryAddSynthesizedType(typeSymbol.GetCciAdapter());
+                typeAdapter = privateImplClass.GetType(typeName)!;
```

Comment:
>! [](http://example.com/codeflow?start=64&length=1)

Do we need this suppression?

Reply:
We're dereferencing `typeAdapter` below.

Reply:
nit: consider doing `Debug.Assert(typeAdapter!.Name == typeName);` instead. That assertion should also ensure the nullable state of `typeAdapter` is not-null.